### ISSUE TITLE
GitHub OAuth authentication with next.js

### ISSUE THREAD

**Issue by gr2m** [Belco90/octoclairvoyant#22]
Hey Mario, thanks again for this great video, I think I understand the problem now:

<a href="https://www.youtube.com/watch?v=gI9N5HmslPw"><img width="960" alt="image" src="https://user-images.githubusercontent.com/39992/98404134-e2bbd380-201e-11eb-8239-49b4726b3624.png"></a>

### Problem statement

- The `octokit` instance is created as part of the App component, which is static, without any backend calls or server-side renderings
- The request to exchange the OAuth code for an OAuth access token happens in the `app/callback` page component. I'm not exactly sure how that works technically, do you have a branch with the WIP code so I can inspect the requests happening to the server in my browser dev tools? I assume that when you open the `/auth/callback` path in your browser, the static parts of the App component render, but before the `/auth/callback` component renders, a request is sent to the backend, where the server code is run, which then responds with a `data` json which is passed to render the page component?
- The problem is that at the point in time when the page component is rendered, the `octokit` instance already exists and there is currently no simple way to set a token for an existing instance. 

### Possible solutions

1. You make the `octokit` instance replaceable, and when you receive a token from `getServerSideProps` you replace the unauthenticated instance with a new one
2. We create a new authentication strategy which similar to [`@octokit/auth-token`](https://github.com/octokit/auth-token.js) strategy but instead of passing a token string, it would accept a callback function, letting you basically define a token getter method.

   ```js
   const octokit = new Octokit({ authStrategy: createCallbackAuth, auth: () => myToken } 
   ```

   I think it would be fairly simple to do and others might have use for it, too. I do not want to build that feature into `@octokit/auth-token` because it's the default authentication strategy, every user of `@octokit/core` or any library built on it (such as `@octokit/rest`) will get the code, whether they use it or not. And bundle size is a critical factor for the JS Octokit SDK. So I would prefer to create a separate authentication strategy
3. We build the `@octokit/auth-oauth-client` SDK as documented at https://github.com/octokit/octokit.js/#oauth-browser-example.

If `1.` is possible, then I'd suggest to do then so you are no longer blocked by it. But I understand (and agree) that it's not an elegant solution.

I'd suggest we collaborate on `2.`, that shouldn't take too long.

Once we have `2.` working, I'd love your input on `3.`. I cannot focus on it because it's not a priority for GitHub, but if there is an Octokit user like yourself you needs it for their work and who is willing to collaborate on it, I'd be happy to find the time on my side.

On thing to take into account for `3.` is that it would take some control over redirects, at least the way it is currently designed. `octokit.auth({ signIn: true })` will redirect to the OAuth Login URL on GitHub. Calling `await { token, isSignedIn, createdAt, scopes } = await auth();` will remove the `?code=...&state=...` from your URL, if present. Do you see any problem with that?

I also haven't considered the case that someone would want to send unauthenticated requests, but I think that would be within the scope of the project. It would simply send unauthenticated requests if the user is not signed in. But it wouldn't handle the use case when you want to redirect the user to the sign in page once they hit the rate limit. I think that is something you will have to handle as part of your app, I'm not sure if it's such a common pattern, we will find out.

I'll looks some more into your app, there are a few apps I want to build myself, but I'm not as experienced with React / next.js. But it is exactly the stack that I'd like to use for my own little side projects, such as https://github.com/gr2m/octokit.rest, https://github.com/octonews/octonews, or the mentioned OAuth App service.

**Comment by Belco90** [comment]
Hey @gr2m! Thank you very much for your help.

I have deployed an update to https://octoclairvoyant.now.sh/ so you can see the GitHub login button within root page in the initial state of the page if you didn't authorize it yet. After authorize it, if you want to get the button back you can just delete the cookie from your browser storage. You'll see something like this:

![image](https://user-images.githubusercontent.com/2677072/98442123-1bcd7400-2103-11eb-88ec-243085280d2e.png)
 
Just bear in mind: the POST request to exchange the code by the access token happens in the SSR so you won't see anything related to that in the network tab of your browser.

### Problem statement
> The octokit instance is created as part of the App component, which is static, without any backend calls or server-side renderings

That's right, but a clarification about it: since I'm using Next.js hybrid mode with SSR, my whole App is instantiated twice: one on SSR and one on CS. I can elaborate more about this if you want me to, but the point about this is that octokit is instantiated twice too. That's why I'm saving the access token in a cookie so when the user access the page, the cookie is sent to my SSR and I can obtain the access token from it to perform any octokit request with it **if necessary**. I'm not using this yet, but I'm planning to do so in the future by sharing links with prefilled filters (so I'd receive the url with some querystrings and then I can get repo and release versions in the SSR, so the user sees already the changelog preloaded when opening the shared url)

> The request to exchange the OAuth code for an OAuth access token happens in the app/callback page component. I'm not exactly sure how that works technically, do you have a branch with the WIP code so I can inspect the requests happening to the server in my browser dev tools? I assume that when you open the /auth/callback path in your browser, the static parts of the App component render, but before the /auth/callback component renders, a request is sent to the backend, where the server code is run, which then responds with a data json which is passed to render the page component?

That's it, you got the behavior: when GitHub come backs to /auth/callback in my app, the request arrives to my Serverless funnction which exchanges the code, then gets the token and build the page itself. This is sent to the Client-Side where at this point you'll see the spinner: that means the access code was obtained successfully so the app is 1) saving it in a cookie and 2) redirecting to root page to start using the app. This last point is the reason why I want to set dynamically `auth` in octokit, or be able to get `auth`dynamically. Anyway as mentioned before, you have octoclairvoyant available to check this workflow.

> The problem is that at the point in time when the page component is rendered, the octokit instance already exists and there is currently no simple way to set a token for an existing instance.

Indeed. This situation won't happen that often because this just occurs when the user is anonymous, auth the app and keep using the SPA dynamically. After the app is auth, every time the user reopen the app, octokit will have corresponding `auth` value since it gets it from cookie, where will be set already for the user.

### Possible solutions
1. The update I deployed already include this, so you can check that as anonymous user the requests don't include the auth header, but then if you auth it on GitHub and search again after the process is completed, the requests include the auth header.
2. This would be fantastic. As I explained in my video, this is how I usually get the token in other apps where I use apollo so I don't set the token but get it dynamically on every request through that callback.
3. It would be amazing to have this solution, but I'm not sure about how the SSR --> CS + hydration process would happen, so there are few things to check first.

As you mentioned, I'm not blocked anymore since I'm using solution 1). If solution 2) it's easy to implement, would be nice to have. But I don't want to push anything to have 3) implemented to be honest. I'm happy to help as much as I can with it tho, but I don't want to take your time for it!

> On thing to take into account for 3. is that it would take some control over redirects, at least the way it is currently designed. octokit.auth({ signIn: true }) will redirect to the OAuth Login URL on GitHub. Calling await { token, isSignedIn, createdAt, scopes } = await auth(); will remove the ?code=...&state=... from your URL, if present. Do you see any problem with that?

I don't see any problem with that. Since I would get the token after awaiting for the sign in (which I can do on the Client-Side), I wouldn't need either /auth/callback page or Serverless function for doing the exchange anymore.

> I also haven't considered the case that someone would want to send unauthenticated requests, but I think that would be within the scope of the project. It would simply send unauthenticated requests if the user is not signed in. But it wouldn't handle the use case when you want to redirect the user to the sign in page once they hit the rate limit. I think that is something you will have to handle as part of your app, I'm not sure if it's such a common pattern, we will find out.

That fits 100% in my app. First version was done using GraphQL GitHub API but people started asking why they had to auth if repo info requested as open. So I switched to REST GitHub API to allow users to use the app anonymously. Handling the rate limit is not an issue: my app handles that already (although I need to improve the mechanism since I have octokit now). You'll see something like this when the limit is reached:

![image](https://user-images.githubusercontent.com/2677072/98442730-c1cead80-2106-11eb-8da1-42b3127e33b3.png)

Let me know if want to know more about this Next.js stack! I love it, and it's the one I'm using for both personal and professional projects lately.

**Comment by gr2m** [comment]
Thanks for sharing all the insights. Let me get `2.` going. I think `3.` is not ideal for next.js, so let's gain some more experience first and maybe we come up with something better down the road.

**Comment by gr2m** [comment]
There you go: https://github.com/octokit/auth-callback.js

To use with `Octokit`

```js
import { createCallbackAuth } = import("@octokit/auth-callback")
const new Octokit({
  authStrategy: createCallbackAuth,
  auth () {
    // return token here
  }
})
```

I haven't tested it myself yet but it's well covered by tests.

**Comment by gr2m** [comment]
ah wait I didn't account for the unauthenticated state, let me add that quickly

**Comment by gr2m** [comment]
there you go: https://github.com/octokit/auth-callback.js/releases/tag/v1.1.0

to check whether the instance is currently authenticated, use

```js
const authentication = await octokit.auth()
```

`authentication.type` will be either `"unauthenticated"` or `"token"`

**Comment by Belco90** [comment]
Awesome! I'm gonna try it into my app and give you some feedback. Thanks!

**Comment by Belco90** [comment]
Sorry, but I think there is something wrong or I can't set it up properly.

> ```javascript
> import { createCallbackAuth } = import("@octokit/auth-callback")
> const new Octokit({
>   authStrategy: createCallbackAuth,
>   auth () {
>     // return token here
>   }
> })
> ```

This is not how is supposed to be configured, is it? I get `Error: [@octokit/auth-callback] Callback passed to createCallbackAuth is not a function` at build time.

Checking the package doc, I should do instead:
```javascript
import { createCallbackAuth } = import("@octokit/auth-callback")

const new Octokit({
  authStrategy: createCallbackAuth(() => /* return token here*/)
})
```

But this way doesn't seem to work fine either. When I perform a request through octokit instance, I get `TypeError: registered.hook is undefined` at run time for both unauthenticated and token types.

I have pushed the latter (which seems the correct one), so you can check both code and platform in #23

**Comment by Belco90** [comment]
@gr2m any update about this? Thanks!

**Comment by gr2m** [comment]
Very sorry Mario, I missed your last comment! Let me what's up

**Comment by gr2m** [comment]
I was able to reproduce the error:
https://runkit.com/gr2m/belco90-octoclairvoyant-22

I'm looking into it

**Comment by gr2m** [comment]
The problem was a change that I made in `@octokit/core` at the same time as I did the new auth strategy and I didn't test the two together. The problem should be resolved with [v2.0.0](https://github.com/octokit/auth-callback.js/releases/tag/v2.0.0), see https://github.com/octokit/auth-callback.js/pull/3

You have to set `auth` to `{ callback () {} }` now

```js
    import { createCallbackAuth } = import("@octokit/auth-callback")
    const new Octokit({
      authStrategy: createCallbackAuth,
      auth { callback: () => "token123'"}
    })
```

**Comment by Belco90** [comment]
Awesome! This is working perfectly now! Thank you very much for your collaboration here 🚀
